src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,298 @@
+#include <absl/strings/match.h>
+#include <absl/strings/str_format.h>
+#include <dlfcn.h>
+#include <sys/ptrace.h>
+#include <sys/wait.h>
+
+#include <csignal>
+#include <string>
+#include <vector>
+
+#include "AccessTraceesMemory.h"
+#include "AllocateInTracee.h"
+#include "ElfUtils/ElfFile.h"
+#include "ElfUtils/LinuxMap.h"
+#include "MachineCode.h"
+#include "OrbitBase/Logging.h"
+#include "UserSpaceInstrumentation/RegisterState.h"
+
+namespace orbit_user_space_instrumentation {
+
+namespace {
+
+using orbit_elf_utils::ElfFile;
+using orbit_elf_utils::ReadModules;
+
+// Size of the small amount of memory we need in the tracee to write machine code into.
+constexpr uint64_t kScratchPadSize = 1024;
+
+// In certain error conditions the tracee is damaged and we don't try to recover from that. We just
+// abort with a fatal log message. None of these errors are expected to occur in operation
+// obvioulsy. That's what the *OrDie methods below are for.
+void FreeMemoryOrDie(pid_t pid, uint64_t address_code, uint64_t size) {
+  auto result = FreeInTracee(pid, address_code, size);
+  FAIL_IF(result.has_error(), "Unable to free previously allocated memory in tracee: \"%s\"",
+          result.error().message());
+}
+
+void RestoreRegistersOrDie(RegisterState& register_state) {
+  auto result = register_state.RestoreRegisters();
+  FAIL_IF(result.has_error(), "Unable to restore register state in tracee: \"%s\"",
+          result.error().message());
+}
+
+uint64_t GetReturnValueOrDie(pid_t pid) {
+  RegisterState return_value_registers;
+  auto result_return_value = return_value_registers.BackupRegisters(pid);
+  FAIL_IF(result_return_value.has_error(), "Unable to read registers after function called :\"%s\"",
+          result_return_value.error().message());
+  return return_value_registers.GetGeneralPurposeRegisters()->x86_64.rax;
+}
+
+// Execute the code at `address_code`. The code segment has to end with an `int3`.
+void ExecuteOrDie(pid_t pid, RegisterState& original_registers, uint64_t address_code) {
+  RegisterState registers_set_rip = original_registers;
+  registers_set_rip.GetGeneralPurposeRegisters()->x86_64.rip = address_code;
+  auto result_restore_registers = registers_set_rip.RestoreRegisters();
+  if (result_restore_registers.has_error()) {
+    FATAL("Unable to continue tracee with PTRACE_CONT.");
+  }
+  if (ptrace(PTRACE_CONT, pid, 0, 0) != 0) {
+    FATAL("Unable to continue tracee with PTRACE_CONT.");
+  }
+  int status = 0;
+  pid_t waited = waitpid(pid, &status, 0);
+  if (waited != pid || !WIFSTOPPED(status) || WSTOPSIG(status) != SIGTRAP) {
+    FATAL("Failed to wait for sigtrap after PTRACE_CONT.");
+  }
+}
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void*> DlopenInTracee(pid_t pid, std::filesystem::path path,
+                                                   uint32_t flag) {
+  // Figure out address of dlopen in libc.
+  OUTCOME_TRY(address_dlopen, FindFunctionAddress(pid, "dlopen", "libc-"));
```

Comment:
interesting, not libdl?

Reply:
As discussed offline: We now try libdl first and use libc as a fallback. I neither is there we return an error.